ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -453,14 +460,33 @@ chainSelectionForBlock
   -> InvalidBlockPunishment m
   -> m (Point blk)
 chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = do
-    (invalid, succsOf, lookupBlockInfo, curChain, tipPoint, ledgerDB)
-      <- atomically $ (,,,,,)
-          <$> (forgetFingerprint <$> readTVar cdbInvalid)
-          <*> VolatileDB.filterByPredecessor  cdbVolatileDB
-          <*> VolatileDB.getBlockInfo         cdbVolatileDB
-          <*> Query.getCurrentChain           cdb
-          <*> Query.getTipPoint               cdb
-          <*> LgrDB.getCurrent                cdbLgrDB
+    (invalid, succsOf', lookupBlockInfo, lookupBlockInfo', curChain, tipPoint, ledgerDB, loeFrag)
+      <- atomically $ do
+          (invalid, succsOf, lookupBlockInfo, curChain, tipPoint, ledgerDB, loeFrag) <-
+                (,,,,,,)
+            <$> (forgetFingerprint <$> readTVar cdbInvalid)
+            <*> VolatileDB.filterByPredecessor  cdbVolatileDB
+            <*> VolatileDB.getBlockInfo         cdbVolatileDB
+            <*> Query.getCurrentChain           cdb
+            <*> Query.getTipPoint               cdb
+            <*> LgrDB.getCurrent                cdbLgrDB
+            <*> readTVar                        cdbLoEFrag
+
+          -- Let these two functions ignore invalid blocks
+          let lookupBlockInfo' = ignoreInvalid    cdb invalid lookupBlockInfo
+              succsOf'         = ignoreInvalidSuc cdb invalid succsOf
+
+              -- Applying the @loeDiff@ would switch from @curChain@ to @loeAnchor@
+              loeFrag' = case AF.intersect curChain loeFrag of
+                -- same as 'cross' from ChainSync client
+                Just (curChainPre, _, _, loeSuf) -> fromJust $ curChainPre `AF.join` loeSuf
+                Nothing                          -> AF.Empty (AF.anchor curChain)
```

Comment:
This branch means that the loe branches off before the immutable tip, right?

My intuition says that should be impossible while the LoE is armed. Maybe it could be violated ephemerally until some harmless race is resolved?

Either way, I suspect this branch deserves a comment discussing this to some extent. Or else a more honest implementation, such as `error "impossible?"`.

Reply:
Yes, I think it could again arise if the governor does not wake up for a relatively long time (not likely, but I think just defaulting to the immutable tip and maybe also having a trace event seem reasonable). Will add a TODO.